### PR TITLE
add top error boundary & reset localStorage on error

### LINKED ISSUE #471: Don't explode if initialization fails
When `localStorage` contains invalid data for whatever reasons (can result due to bugs, or breaking change update), and it throws during initialization, the app should survive it. Simplest solution in that scenario is to clear `localStorage` and use default values.

Currently, if this happens, there's no recourse other than opening devTools and running `localStorage.clear()` + refresh.

### PR DESCRIPTION
**Pull request by dwelle**
fixes #471 

This is not the end state we wanna arrive to, but I've had some issues that I couldn't debug. Maybe as a hotfix it suffices.

### CODE REVIEW

File: src/index.tsx [hunk 1]
```diff
@@ -1371,4 +1371,26 @@ export class App extends React.Component<any, AppState> {
...
+    window.location.reload();
+  }
+
+  static getDerivedStateFromError(error: any) {
+    return { hasError: true };
+  }
+
+  render() {
+    return this.state.hasError ? (
+      <div>Encountered an error. Please reload.</div>
```
Comment (dwelle) [inline src/index.tsx]: When I was debugging, without the `location.reload()` above, it always rendered this and immediately afterwards it rendered the App again (and displayed the error stack). Any ideas why?

Comment (dwelle) [reply src/index.tsx]: My end goal is to display an error message telling the user to reload (I'd render a reload button, for example). And then if reload didn't work, I'd tell the user to confirm that we'll discard his scene (in localStorage). But I couldn't make the error splash screen (the `<div>`) stay on screen without the error being rethrown (rerendered) FSR. No clue why.

**Comment by lipis** [comment]
It should be well thought and display something more than just a text.

**Comment by dwelle** [comment]
> It should be well thought and display something more than just a text.

That's a fallback that won't really be displayed. The browser immediately reloads. But in the long run, it should, yes (see my comments above).

File: src/index.tsx (hunk 1 above)
Comment (dwelle) [reply src/index.tsx]: I think I got this. This is due to CRA dev mode. In prod mode it would work.

**Comment by dwelle** [comment]
@lipis do we merge this? I'm not even sure what's our goal with the error splash screen, but I can work on it in a next PR.

**Comment by lipis** [comment]
I made some small changes, check it out.

**Comment by lipis** [comment]
We could probably check with @vjeux on this one.

File: package.json [hunk 2]
```diff
@@ -41,8 +41,7 @@
   },
   "lint-staged": {
     "*.{js,css,scss,json,md,ts,tsx,html,yml}": [
-      "prettier --write",
-      "git add"
+      "prettier --write"
```
Comment (lipis) [inline package.json]: I removed that due to a warning of the latest `lint-staged` and that caused the package lock update.

**Comment by vjeux** [comment]
This approach makes sense to me. Thanks for the quick fix. We should still try and figure out why this happened in the first place.

**Comment by dwelle** [comment]
> I made some small changes, check it out.

Thanks. I'm actually working on a small UI rewrite, so will force-push.

**Comment by lipis** [comment]
> Thanks. I'm actually working on a small UI rewrite, so will force-push.

Have the message it in the middle of the page.

**Comment by dwelle** [comment]
image

Actually. Maybe don't merge this yet. AFAIK we're deploying CRA still in dev mode. If so, then this solution won't work.

**Comment by dwelle** [comment]
Or *are* we deploying in dev mode, and just bundling in sourceMaps? Because I recall that user errors in prod looked too much like dev mode stack traces.

**Comment by vjeux** [comment]
Could you display the error with stack trace, a textarea with the localstorage content and a link to the github issue tracker?

I’d love to have people reachout with as much debug information when that happens so that we can fix it.

**Comment by dwelle** [comment]
> Could you display the error with stack trace, a textarea with the localstorage content and a link to the github issue tracker?
> 
> I’d love to have people reachout with as much debug information when that happens so that we can fix it.

Alright. In the meanwhile, can someone please check we're deploying in prod mode? I'm getting lost in the GH actions. Where is the CI config we use for build?

**Comment by dwelle** [comment]
Actually, the screen [supplied here](https://user-images.githubusercontent.com/20165741/72793210-16657280-3c32-11ea-8722-14f21a7767a5.png) shows it's a prod build. But the stack trace is nice (apparently mapped). On the other hand, what I'm getting from `error.stack` supplied to the error boundary is completely useless (testing `production` build locally). :thinking:

**Comment by lipis** [comment]
It's being build on Netlify, by running: `npm run build` and published the `build/` directory.

**Comment by dwelle** [comment]
Not sure what the stack trace will look like in practice. We can iterate on this in future PRs. I'd get this merged ASAP.

image

File: src/bug-issue-template.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
+export default `
+# Stack strace
+
+\`\`\`
+// paste stack trace here
+\`\`\`
+
+# localStorage
```
Comment (lipis) [inline src/bug-issue-template.js]:
```suggestion
### localStorage
```

File: src/bug-issue-template.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
+export default `
+# Stack strace
```
Comment (lipis) [inline src/bug-issue-template.js]:
```suggestion
### Stack strace
```

File: src/styles.scss [hunk 5]
```diff
@@ -133,3 +135,49 @@ button {
...
+  }
+
+  .ErrorSplash-details {
+    display: flex;
+    flex-direction: column;
+    align-items: flex-start;
+
+    textarea {
+      width: 100%;
+      margin: 10px 0;
```
Comment (lipis) [inline src/styles.scss]: can we use the `Cascadia` here?

https://github.com/excalidraw/excalidraw/pull/465

File: src/styles.scss [hunk 6]
```diff
@@ -133,3 +135,49 @@ button {
...
+
+  .ErrorSplash-messageContainer {
+    display: flex;
+    flex-direction: column;
+    align-items: center;
+    justify-content: center;
+
+    padding: 40px;
+    background: #fdf4f5;
+    border: 3px solid #d72b3f;
```
Comment (lipis) [inline src/styles.scss]: is this color from: https://yeun.github.io/open-color/

File: src/styles.scss [hunk 7]
```diff
@@ -133,3 +135,49 @@ button {
...
+  justify-content: center;
+
+  .ErrorSplash-messageContainer {
+    display: flex;
+    flex-direction: column;
+    align-items: center;
+    justify-content: center;
+
+    padding: 40px;
+    background: #fdf4f5;
```
Comment (lipis) [inline src/styles.scss]:
```suggestion
    background-color: #fdf4f5;
```

File: src/styles.scss (hunk 5 above)
Comment (dwelle) [reply src/styles.scss]: Tried, is too big for this textareas. Also, no one cares about font in an error message textarea.

Comment (lipis) [reply src/styles.scss]: change the font-size to have it smaller.. I know but for consistency we can have that.. why not?

File: src/styles.scss (hunk 6 above)
Comment (lipis) [reply src/styles.scss]: Maybe use the orange tone instead to have it less prominent...?

Comment (dwelle) [reply src/styles.scss]: It's a critical error. Should be prominent.

**Comment by lipis** [comment]
Let's change it to orange and it's fine by me.

**Comment by dwelle** [comment]
> Let's change it to orange and it's fine by me.

Why should we change it to orange? Let's use orange for warnings, not errors.

Also, why are we hanging this critical hotfix on such petty problems like shades of color?

**Comment by dwelle** [comment]
If you feel like an error splash screen needs restyling, feel free to open new PR. I'm merging this because it affects users right now.

**Pull request merged**